### PR TITLE
Check required functions

### LINKED ISSUE #451: undefined reference to usrsctp_get_timeout
build fails with wrong version of sctplab

expected: build should fail early in configure phase

`usrsctp_get_timeout` is added in https://github.com/sctplab/usrsctp/pull/591

https://github.com/frida/frida-core/blob/ab2581a0b7faea22d87a33c1cc6f2aab8bd88ece/lib/base/p2p-glue.c#L347

error:

```
[57/144] Linking target lib/agent/libfrida-agent.so
FAILED: lib/agent/libfrida-agent.so 
g++  -o lib/agent/libfrida-agent.so ...
ld: lib/base/libfrida-base-1.0.a(p2p-glue.c.o): in function `_frida_sctp_timer_source_get_timeout':
p2p-glue.c:(.text+0x8b3): undefined reference to `usrsctp_get_timeout'
```

### PR DESCRIPTION
**Pull request by milahu**
fix #451 

fail early when required functions are missing

### CODE REVIEW

**Comment by oleavr** [comment]
Thanks! Could you tweak the commit messages slightly:

```
meson: Require usrsctp_get_timeout()
meson: Require Frida fork of GLib
```

(So I can rebase and merge this PR as-is.)

**Comment by milahu** [comment]
fixed. some checks are still missing, but i can make a new PR

File: meson.build [hunk 1]
```diff
@@ -271,7 +282,14 @@ if host_os == 'android'
   backend_deps_private += [libselinux_dep, libsepol_dep]
 endif
 
-tls_provider_dep = dependency('gioopenssl', required: get_option('connectivity'))
+# workaround for https://github.com/mesonbuild/meson/issues/11381
+if get_option('connectivity').auto()
+  error('option "connectivity" cannot be "auto", must be either "enabled" or "disabled"')
+endif
```
Comment (oleavr) [inline meson.build:288]: We should support `auto` , so this would be a regression.

**Review by oleavr — CHANGES_REQUESTED** [review]
Thanks! One concern:

File: meson.build [hunk 2]
```diff
@@ -231,9 +231,20 @@ if get_option('b_sanitize') == 'address'
...
+gio_dep = dependency('gio-2.0')
+
+assert(
+  # vala: Thread.garbage_collect
+  cc.has_function('g_thread_garbage_collect', dependencies: [glib_dep]) and
+  # vala: GLibFork.prepare_to_fork
+  cc.has_function('glib_prepare_to_fork', dependencies: [glib_dep]) and
+  # vala: GIOFork.prepare_to_fork
+  cc.has_function('gio_prepare_to_fork', dependencies: [gio_dep]),
+  'please use vala from https://github.com/frida/vala and glib from https://github.com/frida/glib'
```
Comment (oleavr) [inline meson.build:243]:
```suggestion
  'Please use Vala from https://github.com/frida/vala and GLib from https://github.com/frida/glib'
```

(We could support using vanilla GLib as well, to better integrate with distro packages. That is, for frida-agent/frida-gadget we should use our own fork, but for the rest there's no benefit.)

File: meson.build [hunk 3]
```diff
@@ -271,7 +282,10 @@ if host_os == 'android'
   backend_deps_private += [libselinux_dep, libsepol_dep]
 endif
 
-tls_provider_dep = dependency('gioopenssl', required: get_option('connectivity'))
+tls_provider_dep = dependency('gioopenssl',
+  required: get_option('connectivity'),
+  not_found_message: 'please use glib-networking from https://github.com/frida/glib-networking'
```
Comment (oleavr) [inline meson.build:287]:
```suggestion
  not_found_message: 'Please use GLib Networking from https://github.com/frida/glib-networking'
```

File: meson.build [hunk 4]
```diff
@@ -292,6 +306,11 @@ if nice_dep.found()
                    prefix: '#include <usrsctp.h>')
     cdata.set('HAVE_SCONN_LEN', 1)
   endif
+
+  assert(
+    cc.has_function('usrsctp_get_timeout', dependencies: [usrsctp_dep]),
+    'please use usrsctp with https://github.com/sctplab/usrsctp/pull/591'
```
Comment (oleavr) [inline meson.build:312]:
```suggestion
    'Please use usrsctp with https://github.com/sctplab/usrsctp/pull/591'
```

**Review by oleavr — CHANGES_REQUESTED** [review]
Yay!

**Comment by oleavr** [comment]
Afraid I had to revert this as it caused regressions -- when cross-compiling for iOS the `gio_prepare_to_fork` check fails for some reason. That reminds me we desperately need some CI for this repo :sweat_smile:

**Pull request merged**